src/helpers/strings.ts

```diff
@@ -11,3 +11,75 @@ export const slugify = (str: string): string => {
     .filter(Boolean)
     .join(join);
 };
+
+const millisecond = 1;
+const second = millisecond * 1000;
+const minute = second * 60;
+const hour = minute * 60;
+const day = hour * 24;
+const week = day * 7;
+
+/**
+ * A collection of periods in milliseconds and their suffixes used when creating
+ * time strings.
+ */
+const periods = [
+  ["w", week],
+  ["d", day],
+  ["h", hour],
+  ["m", minute],
+  ["s", second],
+  ["ms", millisecond],
+] as const;
+
+/**
+ * Convert a given `date` to a sleep-compatible time string (e.g. `"1d"` or
+ * `"2h3010s"`).
+ *
+ * Can optionally provide a `now` date to use as the base for the calculation,
+ * otherwise a new date will be created on invocation.
+ */
+export const dateToTimeStr = (
+  /**
+   * The future date to use to convert to a time string.
+   */
+  date: Date,
+
+  /**
+   * Optionally provide a date to use as the base for the calculation.
+   */
+  now = new Date()
+): string => {
+  const isValidDate = !isNaN(date.getTime());
+
+  if (!isValidDate) {
+    throw new Error("Invalid date given to convert to time string");
+  }
+
+  /**
+   * TODO In this eventuality, should we smartly skip the sleep?
+   *
+   * We'd have to return two ops to Inngest, this being skipped and the
+   * next.
+   */
+  if (date <= now) {
+    throw new Error("Cannot sleep until a time in the past");
```

Comment:
@jpwilliams, I am assuming this error will appear in the logs for a given function run in the Inngest dashboard. Could we include the ISO8601 timestamp in the error message? That may help a user determine what bad value in the past that their code may have produced.

Reply:
Or, if this is a timing/race condition, should we warn and then continue with no sleep?

Reply:
Mm I'm not a fan of this `throw` at all!

The executor should probably hold the knowledge here, so perhaps we always push the operation to the executor regardless of when we think the date is, then the executor can call right back if it needs to.